File: scripts/algorithms/S/Split Two Strings to Make Palindrome/Split Two Strings to Make Palindrome.java
// Runtime: 3 ms (Top 100.0%) | Memory: 45.60 MB (Top 14.69%)

class Solution {
    public boolean checkPalindromeFormation(String a, String b) {
        // either way of split should give us a palindrome
        return cut(a, b) || cut(b, a);
    }

    // method to match letters from both ends
    private boolean cut(String a, String b) {
        int i = 0, j = a.length() - 1;
        // converge from both ends till we have same letters
        while (i < j && a.charAt(i) == b.charAt(j)) {
            i++; j--;
        }

        // the case when we surpassed the mid point from both ends
        if (i >= j) return true;
        // the case when there is still a substring left in between
        // or say we didn't reach the mid point
        // we will check if that substring is a palindrome or not
        return isPalindrome(a, i, j) || isPalindrome(b, i, j);
    }

    // method to check if a string is palindrome
    private boolean isPalindrome(String s, int i, int j) {
        while (i < j) {
            if (s.charAt(i++) != s.charAt(j--)) {
                return false;
            }
        }

        return true;
    }
}

File: scripts/algorithms/S/Split Two Strings to Make Palindrome/Split Two Strings to Make Palindrome.cpp
// Runtime: 228 ms (Top 5.18%) | Memory: 36.6 MB (Top 14.81%)
// samll trick: for plaindrome question always try to follow concept that if corners are equal we need to only work
// on middle string to check whether it is also palindrome, instead of check complete strings(both given strings).
class Solution {
public:
    bool ispalind(string x, int i, int j){
        while(i<j){
            if(x[i]!=x[j]) return false;
            i++;
            j--;
        }
        return true;
    }
    bool checkpositions(string a, string b){
        int i=0,j=b.size()-1;
        while(i<j){
            if(a[i]!=b[j]) break;
            i++;
            j--;
        }
         /*
          left cut
          //agar same hote toh
            "ulacfd" ul.zalu //to check for palindrome : za(from b)
            "jizalu" ji.acfd //to check for palindrome : ac(from a)
        */
        /*
         right cut
         //agar samee hote toh
            "ulacfd" jiza.fd //to check for palindrome : za(from b)
            "jizalu" ulac.lu //to check for palindrome : ac(from a)
        */
        return ispalind(a,i,j) || ispalind(b,i,j);
    }
    bool checkPalindromeFormation(string a, string b) {
              //cut one(from left) //cut two(from right)
        return checkpositions(a,b)||checkpositions(b,a);
    }
};